// @amplitude/analytics-react-native/lib/commonjs/plugins/context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Context = void 0;
var _analyticsCore = require("@amplitude/analytics-core");
var _uaParserJs = _interopRequireDefault(require("@amplitude/ua-parser-js"));
var _version = require("../version");
var _reactNative = require("react-native");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const BROWSER_PLATFORM = 'Web';
const IP_ADDRESS = '$remote';
class Context {
  name = '@amplitude/plugin-context-react-native';
  type = 'before';

  // this.config is defined in setup() which will always be called first
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore

  nativeModule = _reactNative.NativeModules.AmplitudeReactNative;
  library = `amplitude-react-native-ts/${_version.VERSION}`;
  constructor() {
    let agent;
    /* istanbul ignore else */
    if (typeof navigator !== 'undefined') {
      agent = navigator.userAgent;
    }
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
    this.uaResult = new _uaParserJs.default(agent).getResult();
  }
  setup(config) {
    this.config = config;
    return Promise.resolve(undefined);
  }
  async execute(context) {
    var _this$nativeModule;
    const time = new Date().getTime();
    const nativeContext = await ((_this$nativeModule = this.nativeModule) === null || _this$nativeModule === void 0 ? void 0 : _this$nativeModule.getApplicationContext(this.config.trackingOptions));
    const appVersion = this.config.appVersion || (nativeContext === null || nativeContext === void 0 ? void 0 : nativeContext.version);
    const platform = (nativeContext === null || nativeContext === void 0 ? void 0 : nativeContext.platform) || BROWSER_PLATFORM;
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment
    const osName = (nativeContext === null || nativeContext === void 0 ? void 0 : nativeContext.osName) || this.uaResult.browser.name;
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment
    const osVersion = (nativeContext === null || nativeContext === void 0 ? void 0 : nativeContext.osVersion) || this.uaResult.browser.version;
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment
    const deviceVendor = (nativeContext === null || nativeContext === void 0 ? void 0 : nativeContext.deviceManufacturer) || this.uaResult.device.vendor;
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment
    const deviceModel = (nativeContext === null || nativeContext === void 0 ? void 0 : nativeContext.deviceModel) || this.uaResult.device.model || this.uaResult.os.name;
    const language = (nativeContext === null || nativeContext === void 0 ? void 0 : nativeContext.language) || (0, _analyticsCore.getLanguage)();
    const country = nativeContext === null || nativeContext === void 0 ? void 0 : nativeContext.country;
    const carrier = nativeContext === null || nativeContext === void 0 ? void 0 : nativeContext.carrier;
    const adid = nativeContext === null || nativeContext === void 0 ? void 0 : nativeContext.adid;
    const appSetId = nativeContext === null || nativeContext === void 0 ? void 0 : nativeContext.appSetId;
    const idfv = nativeContext === null || nativeContext === void 0 ? void 0 : nativeContext.idfv;
    const event = {
      user_id: this.config.userId,
      device_id: this.config.deviceId,
      session_id: this.config.sessionId,
      time,
      ...(appVersion && {
        app_version: appVersion
      }),
      ...(this.config.trackingOptions.platform && {
        platform: platform
      }),
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      ...(this.config.trackingOptions.osName && {
        os_name: osName
      }),
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      ...(this.config.trackingOptions.osVersion && {
        os_version: osVersion
      }),
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      ...(this.config.trackingOptions.deviceManufacturer && {
        device_manufacturer: deviceVendor
      }),
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      ...(this.config.trackingOptions.deviceModel && {
        device_model: deviceModel
      }),
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      ...(this.config.trackingOptions.language && {
        language: language
      }),
      ...(this.config.trackingOptions.country && {
        country: country
      }),
      ...(this.config.trackingOptions.carrier && {
        carrier: carrier
      }),
      ...(this.config.trackingOptions.ipAddress && {
        ip: IP_ADDRESS
      }),
      ...(this.config.trackingOptions.adid && {
        adid: adid
      }),
      ...(this.config.trackingOptions.appSetId && {
        android_app_set_id: appSetId
      }),
      ...(this.config.trackingOptions.idfv && {
        idfv: idfv
      }),
      insert_id: (0, _analyticsCore.UUID)(),
      partner_id: this.config.partnerId,
      plan: this.config.plan,
      ...(this.config.ingestionMetadata && {
        ingestion_metadata: {
          source_name: this.config.ingestionMetadata.sourceName,
          source_version: this.config.ingestionMetadata.sourceVersion
        }
      }),
      ...context,
      library: this.library
    };
    return event;
  }
}
exports.Context = Context;
