// @amplitude/analytics-react-native/lib/commonjs/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AmplitudeReturn", {
  enumerable: true,
  get: function () {
    return _analyticsCore.AmplitudeReturn;
  }
});
Object.defineProperty(exports, "BaseEvent", {
  enumerable: true,
  get: function () {
    return _analyticsCore.BaseEvent;
  }
});
Object.defineProperty(exports, "BeforePlugin", {
  enumerable: true,
  get: function () {
    return _analyticsCore.BeforePlugin;
  }
});
Object.defineProperty(exports, "DestinationPlugin", {
  enumerable: true,
  get: function () {
    return _analyticsCore.DestinationPlugin;
  }
});
Object.defineProperty(exports, "EnrichmentPlugin", {
  enumerable: true,
  get: function () {
    return _analyticsCore.EnrichmentPlugin;
  }
});
Object.defineProperty(exports, "Event", {
  enumerable: true,
  get: function () {
    return _analyticsCore.Event;
  }
});
Object.defineProperty(exports, "EventOptions", {
  enumerable: true,
  get: function () {
    return _analyticsCore.EventOptions;
  }
});
Object.defineProperty(exports, "GroupIdentifyEvent", {
  enumerable: true,
  get: function () {
    return _analyticsCore.GroupIdentifyEvent;
  }
});
Object.defineProperty(exports, "IIdentify", {
  enumerable: true,
  get: function () {
    return _analyticsCore.IIdentify;
  }
});
Object.defineProperty(exports, "ILogger", {
  enumerable: true,
  get: function () {
    return _analyticsCore.ILogger;
  }
});
Object.defineProperty(exports, "IRevenue", {
  enumerable: true,
  get: function () {
    return _analyticsCore.IRevenue;
  }
});
Object.defineProperty(exports, "IdentifyEvent", {
  enumerable: true,
  get: function () {
    return _analyticsCore.IdentifyEvent;
  }
});
Object.defineProperty(exports, "IdentifyOperation", {
  enumerable: true,
  get: function () {
    return _analyticsCore.IdentifyOperation;
  }
});
Object.defineProperty(exports, "IdentityStorageType", {
  enumerable: true,
  get: function () {
    return _analyticsCore.IdentityStorageType;
  }
});
Object.defineProperty(exports, "LogLevel", {
  enumerable: true,
  get: function () {
    return _analyticsCore.LogLevel;
  }
});
Object.defineProperty(exports, "Plugin", {
  enumerable: true,
  get: function () {
    return _analyticsCore.Plugin;
  }
});
Object.defineProperty(exports, "ReactNativeAttributionOptions", {
  enumerable: true,
  get: function () {
    return _analyticsCore.ReactNativeAttributionOptions;
  }
});
Object.defineProperty(exports, "ReactNativeClient", {
  enumerable: true,
  get: function () {
    return _analyticsCore.ReactNativeClient;
  }
});
Object.defineProperty(exports, "ReactNativeConfig", {
  enumerable: true,
  get: function () {
    return _analyticsCore.ReactNativeConfig;
  }
});
Object.defineProperty(exports, "ReactNativeOptions", {
  enumerable: true,
  get: function () {
    return _analyticsCore.ReactNativeOptions;
  }
});
Object.defineProperty(exports, "ReactNativeTrackingOptions", {
  enumerable: true,
  get: function () {
    return _analyticsCore.ReactNativeTrackingOptions;
  }
});
Object.defineProperty(exports, "Result", {
  enumerable: true,
  get: function () {
    return _analyticsCore.Result;
  }
});
Object.defineProperty(exports, "RevenueProperty", {
  enumerable: true,
  get: function () {
    return _analyticsCore.RevenueProperty;
  }
});
Object.defineProperty(exports, "ServerZone", {
  enumerable: true,
  get: function () {
    return _analyticsCore.ServerZone;
  }
});
Object.defineProperty(exports, "ServerZoneType", {
  enumerable: true,
  get: function () {
    return _analyticsCore.ServerZoneType;
  }
});
Object.defineProperty(exports, "SpecialEventType", {
  enumerable: true,
  get: function () {
    return _analyticsCore.SpecialEventType;
  }
});
Object.defineProperty(exports, "Storage", {
  enumerable: true,
  get: function () {
    return _analyticsCore.Storage;
  }
});
Object.defineProperty(exports, "TransportType", {
  enumerable: true,
  get: function () {
    return _analyticsCore.TransportType;
  }
});
var _analyticsCore = require("@amplitude/analytics-core");
